src/libaktualizr/aktualizr_api.h

```diff
@@ -0,0 +1,241 @@
+  /** Download only */
+  AKTUALIZR_UPDATE_DOWNLOAD,
+  /** Install only */
+  AKTUALIZR_UPDATE_INSTALL
+
+} aktualizr_poll_operation_t;
+
+/** Callback type for \ref aktualizr_set_state_change_callback */
+typedef void (*aktualizr_callback_t)(void* clientp, aktualizr_update_state_t);
+
+/// \name Main API
+/// \{
+
+/**
+ * Create a new Aktualizr software updater.
+ * If polling is enabled in the configuration, this will start polling the
+ * update server in the background.
+ */
+aktualizr_t aktualizr_init(aktualizr_config_t configuration);
+
+/**
+ * Shutdown Aktualizr. No other calls to this API may be made after this call.
+ */
+void aktualizr_free(aktualizr_t aktualizr);
+
+/**
+ * Register to be notified when updates are available, downloading or installing.
+ * It is not possible to register to multiple callbacks. Passing NULL as a
+ * callback will disable state change notifications.
+ * \param aktualizr Aktualizr handle
+ * \param callback The callback to call, or 'NULL' to disable notifications
+ * \param userdata Opaque user handle that is passed to the callback
+ */
+void aktualizr_set_state_change_callback(aktualizr_t aktualizr, aktualizr_callback_t callback, void* userdata);
+
+/**
+ * The current state of the update process.
+ */
+aktualizr_update_state_t aktualizr_get_update_state(aktualizr_t aktualizr);
+
+/**
+ * Perform an update cycle.
+ * If Aktualizr isn't currently idle (for example if a polling operation is already in progress) then
+ * an \ref AKTUALIZR_ALREADY_IN_PROGRESS error is returned.
+ */
+aktualizr_res_t aktualizr_run_update(aktualizr_t aktualizr, aktualizr_poll_operation_t operation);
+
+/**
+ * Report system information to the server.
+ * This includes network, hardware and installed package information.
+ * \todo Async? Queued?
+ */
+aktualizr_res_t aktualizr_report_system_information(aktualizr_t aktualizr);
+
+///
+/// \}
+///
+
+///
+/// \name Aktualizr Configuration
+/// \{
+
+/**
+ * Create a new configuration object, which is used to build Aktualizr's
+ * configuration.
+ */
+aktualizr_config_t aktualizr_config_init();
+
+/**
+ * Free an Aktualizr configuration object.
+ * This is needed regardless of whether it is passed to \ref aktualizr_init.
+ */
+void aktualizr_config_free(aktualizr_config_t config);
+
+/**
+ * Load configuration from a file or directory of configuration fragments
+ */
+aktualizr_res_t aktualizr_config_load(aktualizr_config_t config, char* path);
+
+/**
+ * Set log level.
+ * \todo Log to places other than stdout (e.g. systemd journal, callbacks)
+ */
+aktualizr_res_t aktualizr_config_log_level(aktualizr_config_t, int loglevel);
+
+/**
+ * Set a single string configuration option.
+ */
+aktualizr_res_t aktualizr_config_set_string(aktualizr_config_t config, char* section, char* option, char* value);
+
+/**
+ * Set a single integer configuration option.
+ */
+aktualizr_res_t aktualizr_config_set_int(aktualizr_config_t config, char* section, char* option, int value);
```

Comment:
Is this intended for runtime usage? Seems dangerous to change a running system. We currently don't support anything like this. What is the goal?

Reply:
I wanted to make it easy to override small bits of config without either exposing a full set of `aktualizr_config_set_<bla>` operations, or forcing the user to write a temporary file.  The configuration would be baked in after it is handed to `aktualizr_init()`